src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs

```diff
@@ -834,43 +836,81 @@ internal unsafe int GetNonRandomizedHashCode()
             }
         }
 
-        // Use this if and only if 'Denial of Service' attacks are not a concern (i.e. never used for free-form user input),
-        // or are otherwise mitigated
-        internal unsafe int GetNonRandomizedHashCodeOrdinalIgnoreCase()
+        internal int GetNonRandomizedHashCodeOrdinalIgnoreCase()
         {
-            fixed (char* src = &_firstChar)
-            {
-                Debug.Assert(src[this.Length] == '\0', "src[this.Length] == '\\0'");
-                Debug.Assert(((int)src) % 4 == 0, "Managed string should start at 4 bytes boundary");
+            return GetNonRandomizedHashCodeOrdinalIgnoreCaseStatic(ref _firstChar, Length, true);
+        }
 
-                uint hash1 = (5381 << 16) + 5381;
-                uint hash2 = hash1;
+        private static int GetNonRandomizedHashCodeOrdinalIgnoreCaseStatic(ref char firstChar, int length, bool normalizeNonAscii)
+        {
+            Debug.Assert(Unsafe.Add(ref firstChar, length) == '\0', "*(&firstChar + length) == '\\0'");
 
-                uint* ptr = (uint*)src;
-                int length = this.Length;
+            uint hash1 = (5381 << 16) + 5381;
+            uint hash2 = hash1;
 
-                // We "normalize to lowercase" every char by ORing with 0x0020. This casts
-                // a very wide net because it will change, e.g., '^' to '~'. But that should
-                // be ok because we expect this to be very rare in practice.
+            // We "normalize to lowercase" every char by ORing with 0x0020. This casts
+            // a very wide net because it will change, e.g., '^' to '~'. But that should
+            // be ok because we expect this to be very rare in practice.
 
-                const uint NormalizeToLowercase = 0x0020_0020u; // valid both for big-endian and for little-endian
+            const uint NormalizeToLowercase = 0x0020_0020u; // valid both for big-endian and for little-endian
 
-                while (length > 2)
+            int i = 0;
```

Comment:
Nit: if perf ends up degraded, try typing _i_ as _nint_ instead of _int_. TBH I wouldn't bother with this unless the perf test shows regressions and we're looking for low hanging fruit.